### ISSUE TITLE
`copy` with `remote_src` doesn't show diff

### ISSUE THREAD

**Issue by simonLeary42** [ansible/ansible#84309]
### Summary

`copy` with `remote_src` does not show a diff. The [docs](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/copy_module.html) say that the copy module has full diff support, but this doesn't feel like full diff support.

I'd also like to point out my [existing pull request which fixes other missing diff output from the same module which has been neglected for over a year](https://github.com/ansible/ansible/pull/81346)

### Issue Type

Bug Report

### Component Name

copy

### Ansible Version

```console
$ ansible --version
ansible [core 2.17.3]
  config file = /Users/simon/unity/unity-ansible-gitlab/ansible.cfg
  configured module search path = ['/Users/simon/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/simon/.local/lib/python3.12/site-packages/ansible
  ansible collection location = /Users/simon/unity/unity-ansible-gitlab/collections
  executable location = /Users/simon/.local/bin/ansible
  python version = 3.12.7 (main, Oct  1 2024, 02:05:46) [Clang 16.0.0 (clang-1600.0.26.3)] (/Users/simon/.local/bin/python3.12)
  jinja version = 3.1.4
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
ANSIBLE_FORCE_COLOR(my-config-file) = True
ANSIBLE_PIPELINING(my-config-file) = True
CACHE_PLUGIN(my-config-file) = jsonfile
CACHE_PLUGIN_CONNECTION(my-config-file) = $HOME/.cache/ansible
CACHE_PLUGIN_TIMEOUT(my-config-file) = 86400
COLLECTIONS_PATHS(my-config-file) = ['my-ansible-directory/collections']
CONFIG_FILE() = my-config-file
DEFAULT_FORKS(my-config-file) = 200
DEFAULT_GATHERING(my-config-file) = smart
DEFAULT_HOST_LIST(my-config-file) = ['my-ansible-directory/inventory.d']
DEFAULT_MANAGED_STR(my-config-file) = This file is managed by Ansible.%n
DEFAULT_ROLES_PATH(my-config-file) = ['my-ansible-directory/roles']
DIFF_ALWAYS(my-config-file) = True
EDITOR(env: EDITOR) = vim
HOST_KEY_CHECKING(my-config-file) = False

CACHE:
=====

jsonfile:
________
_timeout(my-config-file) = 86400
_uri(my-config-file) = /Users/simon/.cache/ansible

CONNECTION:
==========

local:
_____
pipelining(my-config-file) = True

paramiko_ssh:
____________
host_key_checking(my-config-file) = False
ssh_args(my-config-file) = -o ControlMaster=auto -o ControlPersist=60s

psrp:
____
pipelining(my-config-file) = True

ssh:
___
host_key_checking(my-config-file) = False
pipelining(my-config-file) = True
ssh_args(my-config-file) = -o ControlMaster=auto -o ControlPersist=60s

winrm:
_____
pipelining(my-config-file) = True
```

### OS / Environment

macos 15.0

### Steps to Reproduce

```yaml
- name: play
  hosts: localhost
  gather_facts: false
  tasks:
    - name: remote copy
      ansible.builtin.copy:
        src: /etc/hosts
        dest: /tmp/hosts
        remote_src: true
        mode: "0644"
      diff: true
```

### Expected Results

I expected to see a diff.

### Actual Results

```console
PLAY [play] ************************************************************************************************************************************************

TASK [remote copy] *****************************************************************************************************************************************
changed: [localhost]

PLAY RECAP *************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibot** [comment]
Files identified in the description:

* [`lib/ansible/plugins/action/copy.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/copy.py)
* [`lib/ansible/modules/copy.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/copy.py)

If these files are incorrect, please update the `component name` section of the description or use the [component](https://github.com/ansible/ansibotmini#commands) bot command.

**Comment by simonLeary42** [comment]
I think the problem is with the copy module, not the action plugin: https://github.com/ansible/ansible/blob/c734ac2125f502dcbef2af46b1aba201f850ad05/lib/ansible/plugins/action/copy.py#L456-L458

**Comment by spyinx** [comment]
> I think the problem is with the copy module, not the action plugin:
> 
> https://github.com/ansible/ansible/blob/c734ac2125f502dcbef2af46b1aba201f850ad05/lib/ansible/plugins/action/copy.py#L456-L458

i saw the code, that self._execute_module() method does not return diff data, but _copy_file() method does. the core method _get_diff_data() in ActionBase seems to get diff data between local source file and remote dest file, maybe you can add a method to get diff data of two files which both on remote host

**Comment by s-hertel** [comment]
Duplicate of #66983. Since the source file is remote, it does need to be fixed in lib/ansible/modules/copy.py by returning 'diff' in the result (which should be a list of diff dictionaries, one per modified file). Most of the module_utils used by the copy module already support diff mode, so a large part of the fix will involve threading the diff dictionaries through the `set_*_if_different` AnsibleModule methods it uses.

**Comment by spyinx** [comment]
@simonLeary42  If you want to temporarily fix this issue, it is relatively simple. Without changing the original function of Ansible, you can make the following modifications:
- first: transform `_get_diff_data()` method in `ActionBase` (located in `lib/ansible/plugins/action/__init__.py` )
```python
class ActionBase(ABC):
    # ...

    def _get_remote_data(self, rpath, task_vars):
        """get file data info on remote host"""

        data = {}
        display.debug("Going to peek to see if file has changed permissions")
        peek_result = self._execute_module(
            module_name='ansible.legacy.file', module_args=dict(path=rpath, _diff_peek=True),
            task_vars=task_vars, persist_files=True)
        
        if peek_result.get('failed', False):
            return data, to_text(peek_result.get(u'msg', u''))
        
        if peek_result.get('rc', 0) != 0:
            return None, ''
        
        if peek_result.get('state') in (None, 'absent'):
            data['content'] = u''
        elif peek_result.get('appears_binary'):
            data['binary'] = 1
        elif peek_result.get('size') and C.MAX_FILE_SIZE_FOR_DIFF > 0 and peek_result['size'] > C.MAX_FILE_SIZE_FOR_DIFF:
            data['larger'] = C.MAX_FILE_SIZE_FOR_DIFF
        else:
            display.debug(u"Slurping the file %s" % rpath)
            dest_result = self._execute_module(
                module_name='ansible.legacy.slurp', module_args=dict(path=rpath),
                task_vars=task_vars, persist_files=True)
            if 'content' in dest_result:
                dest_contents = dest_result['content']
                if dest_result['encoding'] == u'base64':
                    dest_contents = base64.b64decode(dest_contents)
                else:
                    raise AnsibleError("unknown encoding in content option, failed: %s" % to_native(dest_result))
                data['header'] = rpath
                data['content'] = to_text(dest_contents)
        
        # Return file data and error msg
        # when ansible run file module failed, error msg can be useful
        return data, ''    
        
    def _get_diff_data(self, destination, source, task_vars, content=None, source_file=True):

        # Note: Since we do not diff the source and destination before we transform from bytes into
        # text the diff between source and destination may not be accurate.  To fix this, we'd need
        # to move the diffing from the callback plugins into here.
        #
        # Example of data which would cause trouble is src_content == b'\xff' and dest_content ==
        # b'\xfe'.  Neither of those are valid utf-8 so both get turned into the replacement
        # character: diff['before'] = u'�' ; diff['after'] = u'�'  When the callback plugin later
        # diffs before and after it shows an empty diff.

        diff = {}

        # Use common method instead
        data, err_msg = self._get_remote_data(destination, task_vars)
        if data is None:
            display.warning(u"Failed to get diff between '%s' and '%s': %s" % (os.path.basename(source), destination, err_msg))
            return diff
        
        if data:
            if 'content' in data:
                diff['before'] = data['content']
            if 'binary' in data:
                diff['dst_binary'] = data['binary']
            if 'larger' in data:
                diff['dst_larger'] = data['larger']
            if 'header' in data:
                diff['before_header'] = data['header']

            if source_file:
                st = os.stat(source)
                if C.MAX_FILE_SIZE_FOR_DIFF > 0 and st[stat.ST_SIZE] > C.MAX_FILE_SIZE_FOR_DIFF:
                    diff['src_larger'] = C.MAX_FILE_SIZE_FOR_DIFF
                else:
                    display.debug("Reading local copy of the file %s" % source)
                    try:
                        with open(source, 'rb') as src:
                            src_contents = src.read()
                    except Exception as e:
                        raise AnsibleError("Unexpected error while reading source (%s) for diff: %s " % (source, to_native(e)))

                    if b"\x00" in src_contents:
                        diff['src_binary'] = 1
                    else:
                        if content:
                            diff['after_header'] = destination
                        else:
                            diff['after_header'] = source
                        diff['after'] = to_text(src_contents)
            else:
                display.debug(u"source of file passed in")
                # If content is not None, then it can the after header name
                diff['after_header'] = content or u'dynamically generated'
                diff['after'] = source

        if self._task.no_log:
            if 'before' in diff:
                diff["before"] = u""
            if 'after' in diff:
                diff["after"] = u" [[ Diff output has been hidden because 'no_log: true' was specified for this result ]]\n"

        return diff
```
> `_get_remote_data()` method simply extracts the code snippet from the original `_get_diff_data()` method that gather dest path's information, resulting in a common function. Finally, it is necessary to modify the `diff ['after_header']`  when `source_file = False`, which defaults to `u'dynamic generated'`. If you want it to have its own name, you can use the content parameter here because the content  under this else branch is useless.
- second: when `remote_src = True`, we can get diff data first before running copy module
```python
# lib/ansible/plugins/action/copy.py

class ActionModule(ActionBase):
    # ...
    
    def run(self, tmp=None, task_vars=None):
        # ...
        
        elif remote_src:
            # Get diff data first
            result['diff'] = []
            data = self._get_remote_data(source, task_vars)
            src_content = u''
            if 'content' in data:
                src_content = data['content']
            if src_content:
                # Only source file exists on remote host, then we need to get diff data
                result['diff'].append(self._get_diff_data(dest, src_content, task_vars, source, False))

            result.update(self._execute_module(module_name='ansible.legacy.copy', task_vars=task_vars))
            return self._ensure_invocation(result)
        # ...
```
finally, Let's run a test:
```shell
[store@spyinx ~]$ cat /tmp/hosts 
xxxxxx

host
[store@spyinx ~]$ cat /etc/hosts
127.0.0.1	localhost	localhost.localdomain	localhost4	localhost4.localdomain4
::1	localhost	localhost.localdomain	localhost6	localhost6.localdomain6


172.20.240.136	spyinx

[store@spyinx ~]$ cat test_ansible_diff.yml 
- name: play
  hosts: localhost
  gather_facts: false
  tasks:
    - name: remote copy
      copy:
        src: /etc/hosts
        dest: /tmp/hosts
        remote_src: true
        mode: "0644"
      diff: true
[store@spyinx ~]$ ansible-playbook -i hosts test_ansible_diff.yml 

PLAY [play] ************************************************************************************************************************************************************

TASK [remote copy] *****************************************************************************************************************************************************
--- before: /tmp/hosts
+++ after: /etc/hosts
@@ -1,3 +1,6 @@
-xxxxxx
+127.0.0.1	localhost	localhost.localdomain	localhost4	localhost4.localdomain4
+::1	localhost	localhost.localdomain	localhost6	localhost6.localdomain6
 
-host
+
+172.20.240.136	spyinx
+

changed: [localhost]

PLAY RECAP *************************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

[store@spyinx ~]$ 
```
my changes have not consider some special situations, so it may occur some error, you can try it and give me some advices, thanks

**Comment by s-hertel** [comment]
Thanks for finding, reporting and fixing bugs :smiley: I'm going to close this one so it can be tracked in one place: https://github.com/ansible/ansible/issues/66983.

**Comment by simonLeary42** [comment]
I have also concocted a workaround worthy of a thumbs down:
```yaml
- name: generate diff manually since ansible copy module has bug
  ansible.builtin.command:
    argv: [python]
    stdin: |
      import json
      with open("{{ dest }}", "r", encoding="utf8") as fp:
          before = fp.read()
      with open("{{ src }}", "r", encoding="utf8") as fp:
          after = fp.read()
      print(json.dumps({"before": before, "after": after}))
  register: manual_diff
  check_mode: false
  changed_when: false

- name: dump manual diff as ansible task result
  return_args:
    # changed must be true in order for diff to be printed
    changed: "{{ (manual_diff.stdout | from_json).before != (manual_diff.stdout | from_json).after }}"
    diff: "{{ manual_diff.stdout | from_json }}"

- name: copy
  ansible.builtin.copy:
    src: "{{ src }}"
    dest: "{{ dest }}"
    remote_src: true
```

this is using my own action plugin `return_args`:
```py
# callback plugin debugging tool
from ansible.plugins.action import ActionBase
class ActionModule(ActionBase):
    def run(self, tmp=None, task_vars=None):
        return self._task.args
```

**Comment by s-hertel** [comment]
A workaround is fine. Tracking the diff where it's made is just easier than double book-keeping, and the thumbs down is intended to save new contributors from heading down the wrong path blindly.

Here's a rough draft adding diff mode support to the module, all feedback on the shape of the diff is welcome (I just kind of threw stuff at the wall): https://github.com/ansible/ansible/pull/84324.
